### PR TITLE
Qc merge

### PR DESCRIPTION
**Pull request by DouglasAbrams**

### CODE REVIEW

File: single_cell/qc.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+    workflow.setobj(
+        obj=mgd.OutputChunks('pseudobulk_group'),
+        value=list(pseudobulk_groups),
+    )
+
+    workflow.subworkflow(
+        name='pseudobulk_group_level_qc_workflow',
+        ctx={'mem':450, 'docker_image': config['docker']['single_cell_pipeline']},
```
Comment (diljotgrewal) [inline single_cell/qc.py:45]: fix.

File: single_cell/qc.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        func="single_cell.qc.pseudobulk_group_level_qc_workflow",
+        axes=('pseudobulk_group',),
+        args=(
+            mgd.InputFile('pseudobulk_group', 'pseudobulk_group', fnames={psg: psg for psg in pseudobulk_groups}),
```
Comment (diljotgrewal) [inline single_cell/qc.py:49]: remove dict comprehension

File: single_cell/qc.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+        ),
+    )
+ 
+    return workflow
+
+def pseudobulk_group_level_qc_workflow(pseudobulk_group, data,  mutationreport, grouplevelmaf, 
+                                  grouplevel_high_impact_maf, grouplevel_snvs, grouplevel_high_impact_merged_snvs, config, out_dir, tmp_dir):
+    sampledata = data[pseudobulk_group]
+    samplegroups = sampledata.keys()
```
Comment (diljotgrewal) [inline single_cell/qc.py:68]: keep line length <90

File: single_cell/qc.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            mgd.InputFile('hmmcopy_reads', 'sample_id', 'library_id', fnames=hmmcopy_reads),
+            mgd.InputFile('hmmcopy_segs', 'sample_id', 'library_id', fnames=hmmcopy_segs),
+            mgd.InputFile('hmmcopy_metrics', 'sample_id', 'library_id', fnames=hmmcopy_metrics),
+            mgd.InputFile('alignment_metrics', 'sample_id', 'library_id', fnames=alignment_metrics),
+            mgd.InputFile('gc_metrics', 'sample_id', 'library_id', fnames=gc_metrics),       
+            mgd.InputFile('indel_files', 'sample_id', 'library_id', fnames=indel_files),       
+            mgd.OutputFile('sample_level_report_htmls', 'sample_id', 'library_id', fnames=sample_level_report_htmls),
+            mgd.OutputFile('mafs', 'sample_id', 'library_id', fnames=mafs),
+            mgd.OutputFile('snvs_all', 'sample_id', 'library_id', fnames=snvs_all),
+            tmp_dir,
```
Comment (diljotgrewal) [inline single_cell/qc.py:136]: remove these dirs.

File: single_cell/workflows/qc/__init__.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    datatype_summary_csv =  os.path.join(prefix,  "datatype_summary.csv")
+    
+    #hardcoded for now
+    vepdata = "/work/shah/reference/vep/"
+    genomeref = "/work/shah/reference/genomes/GRCh37-lite/GRCh37-lite.fa"
+
+    workflow = pypeliner.workflow.Workflow(ctx=ctx)
+    
+
+    # if len(indelvcfs) > 1:
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:99]: remove

File: single_cell/workflows/qc/__init__.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            mgd.OutputFile(CNplot_png), 
+            mgd.OutputFile(datatype_summary_csv),
+
+        ),
+    )
+
+    mutations_per_cell_png =  os.path.join(outprefix, "mutations_per_cell.png")
+    summary_csv =  os.path.join(outprefix, "summary.csv")
+    snvs_high_impact_csv =  os.path.join(outprefix, "snvs_high_impact.csv")
+    # snvs_all_csv =  os.path.join(outprefix, "snvs_all.csv")
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:182]: remove

File: single_cell/workflows/qc/input.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/input.yaml:1]: remove

File: single_cell/qc.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            mgd.InputFile('outpaths', 'sample_id', 'library_id', fnames=outpaths),
+        ),
+    )
+    
+    workflow.subworkflow(
+        name='create_pseudobulk_group_workflow',
+        ctx={'mem':450, 'docker_image': config['docker']['single_cell_pipeline']},
+        func="single_cell.workflows.qc.create_pseudobulk_group_workflow",
+        args=(
+            pseudobulk_group,
```
Comment (diljotgrewal) [inline single_cell/qc.py:147]: what is this? a dict of filenames? dont pass it like this.

File: single_cell/workflows/qc/__init__.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        args=(
+            mgd.InputFile(merged_maf),
+            mgd.OutputFile(high_impact_maf),
+        ),
+    )
+    workflow.transform(
+        name='merge_snvs',
+        func='single_cell.workflows.qc.tasks.merge_snvs',
+        args=(
+            sample_all_snv_csvs,
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:36]: is this an input file?

File: single_cell/workflows/qc/__init__.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            mgd.InputFile(merged_snvs),
+            mgd.OutputFile(merged_high_impact_snvs),
+        )
+    )
+
+    workflow.transform(
+        name='mutationreport',
+        func='single_cell.workflows.qc.tasks.create_mutation_report',
+        args=(
+            pseudobulk_group,
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:54]: fix.

File: single_cell/workflows/qc/__init__.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+def create_sample_level_plots(cell_id, library_id, mappability_file, 
+    strelka_file, museq_file, cosmic_status_file, snpeff_file, dbsnp_status_file, trinuc_file, 
+    counts_file, breakpoint_annotation, breakpoint_counts, haplotype_allele_data, annotation_metrics,
+    hmmcopy_reads, hmmcopy_segs, hmmcopy_metrics, alignment_metrics, gc_metrics, indel_file, reporthtml, maf, 
+    snvs_all_csv, tmp_dir, out_dir, outpath
+):
+
+    ctx = {'mem_retry_increment': 2, 'disk_retry_increment': 50, 'ncpus': 1, }
+
+    tantalus = "/work/shah/tantalus/"
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:74]: remove this?

File: single_cell/workflows/qc/__init__.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    summary_csv =  os.path.join(prefix, "summary.csv")
+    snvs_high_impact_csv =  os.path.join(prefix, "snvs_high_impact.csv")
+    trinuc_csv =  os.path.join(prefix, "trinuc.csv")
+    snv_adjacent_distance_png =  os.path.join(prefix, "snv_adjacent_distance.png")
+    snv_genome_count_png =  os.path.join(prefix,  "snv_genome_count.png")  
+    snv_cell_counts_png =  os.path.join(prefix,  "snv_cell_counts.png")  
+    snv_alt_counts_png =  os.path.join(prefix,  "snv_alt_counts.png")  
+    rearranegementtype_distribution_png =  os.path.join(prefix,  "rearranegementtype_distribution.png")  
+    chromosome_types_png =  os.path.join(prefix,  "chromosome_types.png") 
+    BAFplot_png =  os.path.join(prefix,  "BAFplot.png")
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:88]: i dont think this is PEP8 compliant.

File: single_cell/workflows/qc/__init__.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    snv_cell_counts_png =  os.path.join(prefix,  "snv_cell_counts.png")  
+    snv_alt_counts_png =  os.path.join(prefix,  "snv_alt_counts.png")  
+    rearranegementtype_distribution_png =  os.path.join(prefix,  "rearranegementtype_distribution.png")  
+    chromosome_types_png =  os.path.join(prefix,  "chromosome_types.png") 
+    BAFplot_png =  os.path.join(prefix,  "BAFplot.png") 
+    CNplot_png =  os.path.join(prefix,  "CNplot.png") 
+    datatype_summary_csv =  os.path.join(prefix,  "datatype_summary.csv")
+    
+    #hardcoded for now
+    vepdata = "/work/shah/reference/vep/"
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:93]: remove these?

File: single_cell/workflows/qc/__init__.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+            mgd.InputFile(breakpoint_counts),      
+            mgd.InputFile(haplotype_allele_data),     
+            mgd.InputFile(annotation_metrics),  
+            mgd.InputFile(hmmcopy_reads),  
+            mgd.InputFile(hmmcopy_segs),  
+            mgd.InputFile(hmmcopy_metrics),  
+            mgd.InputFile(alignment_metrics),  
+            mgd.InputFile(gc_metrics),  
+            library_id,
+            tantalus,
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:158]: need to make it tantalus independent?

File: single_cell/workflows/qc/__init__.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+    snvs_high_impact_csv =  os.path.join(outprefix, "snvs_high_impact.csv")
+    # snvs_all_csv =  os.path.join(outprefix, "snvs_all.csv")
+    trinuc_csv =  os.path.join(outprefix, "trinuc.csv")
+    snv_adjacent_distance_png =  os.path.join(outprefix, "snv_adjacent_distance.png")
+    snv_genome_count_png =  os.path.join(outprefix,  "snv_genome_count.png")  
+    snv_cell_counts_png =  os.path.join(outprefix,  "snv_cell_counts.png")  
+    snv_alt_counts_png =  os.path.join(outprefix,  "snv_alt_counts.png")  
+    rearranegementtype_distribution_png =  os.path.join(outprefix,  "rearranegementtype_distribution.png")  
+    chromosome_types_png =  os.path.join(outprefix,  "chromosome_types.png") 
+    BAFplot_png =  os.path.join(outprefix,  "BAFplot.png")
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:190]: fix variable naming

File: single_cell/workflows/qc/tasks.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
+import os
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:1]: lint this file please. spacing between functions is inconsistent.

File: single_cell/workflows/qc/tasks.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    dir_len = len(dir.split("/"))
+    filenames =  sorted(glob.glob(dir + "/*/*/*snv*all*csv", recursive=True))
+    for filename in filenames:
+
+        name_len = len(filename.split("/"))
+        samplelabel = filename.split("/")[dir_len : name_len - 1]
+        output[tuple(samplelabel)] = filename
+    return output
+
+def gunzip_file(infile, outfile):
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:71]: see if this exists in single_cell.utils.helpers. 

if it doesnt, move it there.

File: single_cell/workflows/qc/tasks.py [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+def _get_indelvcfs(dir):
+    indel_vcfs = []
+    indel_vcfs += sorted(glob.glob(dir + "/results/variants/" + "*_strelka_indel.vcf.gz" ))
+    indel_vcfs += sorted(glob.glob(dir + "/results/" + "*_strelka_indel.vcf.gz" ))
+    indel_vcfs += sorted(glob.glob(dir + "/results/variant_calling/*/" + "*indel.vcf.gz" ))
+    return indel_vcfs
+
+#dummie outputs for testing
+def dummie(files):
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:54]: do you need this in prod?

File: single_cell/workflows/qc/tasks.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+    if vcf_file.endswith('.gz'):
+        helpers.makedirs(tempdir)
+        vcf_unzipped = os.path.join(tempdir, 'unzipped_vcf.vcf')
+        gunzip_file(vcf_file, vcf_unzipped)
+    else:
+        vcf_unzipped = vcf_file
+
+    cmd = ["vcf2maf.pl", "--input-vcf", vcf_unzipped, "--output-maf", output_maf,
+           "--ref-fasta", reference, "--filter-vcf", "0", "--vep-path", "/home/abramsd/miniconda3/envs/r-environment/bin",
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:86]: fix vep path

File: single_cell/workflows/qc/tasks.py [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+        vcf_unzipped = os.path.join(tempdir, 'unzipped_vcf.vcf')
+        gunzip_file(vcf_file, vcf_unzipped)
+    else:
+        vcf_unzipped = vcf_file
+
+    cmd = ["vcf2maf.pl", "--input-vcf", vcf_unzipped, "--output-maf", output_maf,
+           "--ref-fasta", reference, "--filter-vcf", "0", "--vep-path", "/home/abramsd/miniconda3/envs/r-environment/bin", 
+           "--vep-data", vepdata]
+
+    pypeliner.commandline.execute(*cmd, docker_image="shub://rdmorin/cancer_docker_singularity:vcf2maf")
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:89]: docker image needs to come from the config.

File: single_cell/workflows/qc/tasks.py [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def sample_level_report(mutations_per_cell, summary, 
+    snvs_high_impact, snvs_all, trinuc, snv_adjacent_distance, snv_genome_count, 
+    snv_cell_counts, snv_alt_counts, rearranegementtype_distribution, chromosome_types,
+    BAFplot, CNplot, datatype_summary, maf, html_file, out_dir, sample_id
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:95]: fix variable naming.

File: single_cell/run.py [hunk 22]
```diff
@@ -17,11 +17,12 @@
 from single_cell.split_bam import split_bam_pipeline
 from single_cell.sv_genotyping import sv_genotyping_pipeline
 from single_cell.variant_calling import variant_calling_pipeline
+from single_cell.qc import qc_pipeline
 
 
 def main():
     args = parse_args()
-
+
```
Comment (diljotgrewal) [inline single_cell/run.py:25]: pep-8

File: single_cell/workflows/qc/scripts/scgenome-analysis.py [hunk 23]
```diff
@@ -0,0 +1,429 @@
+def scgenome_analysis(sample_id,  mappability_file, strelka_file, 
+    museq_file, cosmic_status_file, snpeff_file, dbsnp_status_file, trinuc_file, counts_file,
+    breakpoint_annotation, breakpoint_count, haplotype_allele_data, annotation_metrics,
+    hmmcopy_reads, hmmcopy_segs, hmmcopy_metrics, alignment_metrics, gc_metrics,
+    library_id, dir, prefix, outprefix, mutations_per_cell, summary, 
+    snvs_high_impact, snvs_all, trinuc, snv_adjacent_distance, snv_genome_count, 
+    snv_cell_counts, snv_alt_counts, rearranegementtype_distribution, chromosome_types,
+    BAFplot, CNplot, datatype_summary
+):
+    import matplotlib
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/scripts/scgenome-analysis.py:10]: dont import inside functions

File: single_cell/qc.py [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+        value=list(patients),
+    )
+
+    workflow.subworkflow(
+        name='patient_level_qc_workflow',
+        func="single_cell.qc.patient_level_qc_workflow",
+        axes=('patient',),
+        args=(
+            mgd.InputFile('patient', 'patient', fnames=patient_names),
+            data,
```
Comment (diljotgrewal) [inline single_cell/qc.py:58]: what's data??

File: single_cell/qc.py [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+    gc_metrics =  {samplegroup: data["gc_metrics"] 
+        for samplegroup, data in sampledata.items()
+    }
+    indel_files =  {samplegroup: data["indel_file"] 
+        for samplegroup, data in sampledata.items()
+    }
+    vep_refdir = config['vep_reference_dir']
+
+
+    sample_level_report_htmls =  {samplegroup: os.path.join(out_dir, patient, samplegroup[0], samplegroup[1], "mainreport.html")
```
Comment (diljotgrewal) [inline single_cell/qc.py:144]: line too long.

File: single_cell/qc.py [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+            mgd.InputFile('outpaths', 'sample_id', 'library_id', fnames=outpaths),
+            vep_refdir
+        ),
+    )
+    
+    workflow.subworkflow(
+        name='create_patient_workflow',
+        func="single_cell.workflows.qc.create_patient_workflow",
+        args=(
+            patient,
```
Comment (diljotgrewal) [inline single_cell/qc.py:203]: is this a file??

File: single_cell/utils/inpututils.py [hunk 27]
```diff
@@ -166,6 +166,17 @@ def load_yaml(path):
...
     return data
 
+def load_qc_input(path):
+    data= {}
+    yaml = load_yaml(path)
+    for pseudobulk_group, pseudobulk_group_data in yaml.items():
+        data[pseudobulk_group] = {}
+        for sample, sample_data in pseudobulk_group_data.items():
+            for library, library_data in sample_data.items():
+                data[pseudobulk_group][(sample, library)] = {data_label: data for data_label, data in library_data.items()}
```
Comment (diljotgrewal) [inline single_cell/utils/inpututils.py:176]: line too long.

File: single_cell/workflows/qc/__init__.py [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+            mgd.OutputFile(snv_adjacent_distance_png), 
+            mgd.OutputFile(snv_genome_count_png), 
+            mgd.OutputFile(snv_cell_counts_png), 
+            mgd.OutputFile(snv_alt_counts_png), 
+            mgd.OutputFile(rearranegementtype_distribution_png), 
+            mgd.OutputFile(chromosome_types_png),
+            mgd.OutputFile(baf_plot_png), 
+            mgd.OutputFile(cn_plot_png), 
+            mgd.OutputFile(datatype_summary_csv),
+
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/__init__.py:145]: remove this line.

File: single_cell/workflows/qc/metadata.yaml [hunk 29]
```diff
@@ -0,0 +1,5 @@
+metadata: "metadata/fitness_pseudobulk_march17.csv"
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/metadata.yaml:1]: delete?

File: single_cell/workflows/qc/scripts/single_cell_qc_plots.py [hunk 30]
```diff
@@ -0,0 +1,459 @@
...
+    return filtered_data[0], filtered_data[1]
+
+
+def plot_mutations_per_cell(snv_data, snv_count_data, mutations_per_cell, prefix):
+
+    allcells = snv_count_data.loc[:, ["cell_id"]].drop_duplicates()
+    run_bulk_snv_analysis(snv_data, snv_count_data, allcells, results_prefix=prefix)
+
+    #Plot number of mutations per cell
+    percell = snv_count_data[snv_count_data.alt_counts > 0].groupby('cell_id').size().reset_index(name='Number of mutations')
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/scripts/single_cell_qc_plots.py:88]: line too long.

File: single_cell/workflows/qc/scripts/single_cell_qc_plots.py [hunk 31]
```diff
@@ -0,0 +1,459 @@
...
+    nmuts = len(snv_data[["chrom", "coord", "ref", "alt"]].drop_duplicates()["coord"])
+    ncells = len(snv_count_data.cell_id.unique())
+    data = {'ncells':[ncells], 'nmutations':[nmuts]}
+    df = pd.DataFrame(data)
+    df.to_csv(summary)
+
+
+def write_high_impact_snvs(snv_data, snv_count_data, snvs_high_impact, snvs_all):
+    # Get high impact mutations and write to file
+    high_impact = (snv_data.query('effect_impact == "HIGH" | effect == "NON_SYNONYMOUS_CODING" | effect == "STOP_GAINED"')
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/scripts/single_cell_qc_plots.py:107]: line length.

File: single_cell/workflows/qc/scripts/single_cell_qc_plots.py [hunk 32]
```diff
@@ -0,0 +1,459 @@
...
+
+    return allele_data
+
+
+def plotbaf(allele_data, baf_plot):
+    merged_allele_data = (
+        allele_data.groupby(['chromosome', 'hap_label', 'end', 'start'])[['allele_1', 'allele_2', 'total_counts']]
+        .aggregate({'allele_1': sum, 'allele_2': sum, 'total_counts': sum}).reset_index())
+
+    merged_allele_data['baf'] = np.minimum(merged_allele_data['allele_1'], merged_allele_data['allele_2']) / merged_allele_data['total_counts'].astype(float)
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/scripts/single_cell_qc_plots.py:306]: line length.

File: single_cell/workflows/qc/tasks.py [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+
+    if vcf_file.endswith('.gz'):
+        helpers.makedirs(tempdir)
+        vcf_unzipped = os.path.join(tempdir, 'unzipped_vcf.vcf')
+        helpers.gunzip_file(vcf_file, vcf_unzipped)
+    else:
+        vcf_unzipped = vcf_file
+
+    cmd = [
+        '/opt/local/singularity/3.6.0/bin/singularity', 'run', '--bind', '/juno/work/shah', 'docker://docker.io/wgspipeline/vcf2maf:v0.0.1',
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:78]: remove this line.

File: single_cell/workflows/qc/tasks.py [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+        'vcf2maf.pl', '--input-vcf', vcf_unzipped, '--output-maf', output_maf,
+        '--vep-path', '/usr/local/bin',
+        '--ref-fasta',
+        os.path.join(reference, 'homo_sapiens', '99_GRCh37', 'Homo_sapiens.GRCh37.75.dna.primary_assembly.fa.gz'),
+        '--filter-vcf', os.path.join(reference, 'ExAC_nonTCGA.r0.3.1.sites.vep.vcf.gz'),
+        '--vep-data', reference,
+    ]
+    cmd = ' '.join(map(str,cmd))
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:86]: replace with pypeliner.commandline.execute

File: single_cell/workflows/qc/tasks.py [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+    filenames =  sorted(glob.glob(dir + "/*/*/*snv*all*csv", recursive=True))
+    for filename in filenames:
+
+        name_len = len(filename.split("/"))
+        samplelabel = filename.split("/")[dir_len : name_len - 1]
+        output[tuple(samplelabel)] = filename
+    return output
+
+
+def vcf2maf(vcf_file, output_maf, tempdir, reference):
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:68]: pass a docker container in here

File: single_cell/qc.py (hunk 24 above)
Comment (DouglasAbrams) [reply single_cell/qc.py:58]: data = inpututils.load_qc_input(args["input_yaml"])

then inside the function,
    sampledata = data[patient]

File: single_cell/qc.py (hunk 26 above)
Comment (DouglasAbrams) [reply single_cell/qc.py:203]: no

File: single_cell/qc.py (hunk 24 above)
Comment (diljotgrewal) [reply single_cell/qc.py:58]: right, so this contains file paths. 


you need to wrap the filepaths as managed files.

File: single_cell/qc.py (hunk 26 above)
Comment (diljotgrewal) [reply single_cell/qc.py:202]: see line 57. i think this is a file.

File: single_cell/run.py (hunk 22 above)
Comment (diljotgrewal) [reply single_cell/run.py:25]: @DouglasAbrams

File: single_cell/workflows/qc/__init__.py (hunk 9 above)
Comment (diljotgrewal) [reply single_cell/workflows/qc/__init__.py:39]: @DouglasAbrams

File: single_cell/workflows/qc/tasks.py [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+    cmd = ['/opt/local/singularity/3.6.0/bin/singularity', 'run', '--bind', '/juno/work/shah', '{}'.format(docker_image), 
+        'vcf2maf.pl', '--input-vcf', vcf_unzipped, '--output-maf', output_maf,
+        '--vep-path', '/usr/local/bin',
+        '--ref-fasta',
+        os.path.join(reference, 'homo_sapiens', '99_GRCh37', 'Homo_sapiens.GRCh37.75.dna.primary_assembly.fa.gz'),
+        '--filter-vcf', os.path.join(reference, 'ExAC_nonTCGA.r0.3.1.sites.vep.vcf.gz'),
+        '--vep-data', reference,
+    ]
+
+    pypeliner.commandline.execute(*cmd)
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:86]: pypeliner.commandline.execute(*cmd, docker_image=docker_image)

File: single_cell/workflows/qc/tasks.py [hunk 37]
```diff
@@ -0,0 +1,109 @@
...
+def vcf2maf(vcf_file, output_maf, tempdir, reference, docker_image):
+
+    if vcf_file.endswith('.gz'):
+        helpers.makedirs(tempdir)
+        vcf_unzipped = os.path.join(tempdir, 'unzipped_vcf.vcf')
+        helpers.gunzip_file(vcf_file, vcf_unzipped)
+    else:
+        vcf_unzipped = vcf_file
+
+    cmd = ['/opt/local/singularity/3.6.0/bin/singularity', 'run', '--bind', '/juno/work/shah', '{}'.format(docker_image),
```
Comment (diljotgrewal) [inline single_cell/workflows/qc/tasks.py:77]: remove all singularity args here.

File: single_cell/qc.py (hunk 26 above)
Comment (DouglasAbrams) [reply single_cell/qc.py:203]: it isnt a file, its a dictionary of strings - i just didnt know how else to do it

File: single_cell/qc.py (hunk 24 above)
Comment (DouglasAbrams) [reply single_cell/qc.py:58]: it contains sample-specific paths. its passed in where i am working at the patient level. How do i do that?

File: single_cell/workflows/qc/tasks.py (hunk 36 above)
Comment (DouglasAbrams) [reply single_cell/workflows/qc/tasks.py:86]: I don't think the command will work if i do it

Comment (DouglasAbrams) [reply single_cell/workflows/qc/tasks.py:86]: should i add it both to the cmd and as an argument to the pypeliner call?

File: single_cell/workflows/qc/tasks.py (hunk 37 above)
Comment (DouglasAbrams) [reply single_cell/workflows/qc/tasks.py:77]: I'm just doing this temporarily until this all gets built into a container and I can test i using docker containers. I need it like this to be able to test.

File: single_cell/qc.py [hunk 38]
```diff
@@ -0,0 +1,222 @@
...
+        obj=mgd.OutputChunks('patient'),
+        value=list(patients),
+    )
+
+    workflow.subworkflow(
+        name='patient_level_qc_workflow',
+        func="single_cell.qc.patient_level_qc_workflow",
+        axes=('patient',),
+        args=(
+            mgd.InputFile('patient', 'patient', fnames=patient_names),
```
Comment (diljotgrewal) [inline single_cell/qc.py:57]: mgd.InputInstance('patient')

**Pull request merged**